### PR TITLE
Report back errors to client for DeleteUnpubIData

### LINKED ISSUE #807: Elders assume for IData delete requests that Adults always report success
This scenario is specific to immutable data as it's the only one where Elders delegate to Adults (and sometimes Elders). For phase 1 the Elder effectively delegates to itself, but the flow is generic in anticipation for future phases.

In `idata_holder.rs` is the logic run by Adults (and Elders) when putting and deleting immutable data. Currently that can return things like `NdError::AccessDenied` if you try to delete data you're not the owner of. However in `idata_handler.rs` the Elder always assume that mutation requests when delegated to Adults (and Elders) succeed.

See 
`IDataHandler::handle_put_idata_resp`
`IDataHandler::handle_delete_unpub_idata_resp`
where the `NdResult` is unused.

For Put requests this probably doesn't have a too big impact, but **for deletions this mean that you can delete other users `UnpublishedIData`**. The integration test for this is in `delete_immutable_data()` where it is disabled
```rust
    // Currently Elders always assume mutation requests (put/delete) to Adults succeed.
    if false {
        // Delete unpublished data without being the owner
        common::send_request_expect_err(
            &mut env,
            &mut client_b,
            Request::DeleteUnpubIData(IDataAddress::Unpub(unpub_idata_address)),
            NdError::AccessDenied,
        );
    }

```

### PR DESCRIPTION
**Pull request by octol**
Resolves #807

### CODE REVIEW

File: src/data_handler/idata_op.rs [hunk 1]
```diff
@@ -92,9 +94,27 @@ impl IDataOp {
...
+            })
+            .filter(|(_, response)| {
+                if let Response::Mutation(Err(_)) = response {
+                    true
+                } else {
+                    false
+                }
+            })
+            .collect()
+    }
```
Comment (octol) [inline src/data_handler/idata_op.rs]: This function seems clunky, must be a better way?

File: src/data_handler/idata_handler.rs [hunk 2]
```diff
@@ -286,46 +286,64 @@ impl IDataHandler {
...
-            warn!(
-                "{}: Failed to get metadata from DB: {:?}",
-                self, idata_address
-            );
-            None
-        });
-
-        if let Some(mut metadata) = metadata {
-            if !metadata.holders.remove(&sender) {
+        // TODO - Only rudimentary checks for if requests to Adult nodes were successful. These
```
Comment (jeanphilippeD) [inline src/data_handler/idata_handler.rs]: Function is a bit on the bigger side, but I guess we can always refactor later.

File: src/data_handler/idata_op.rs (hunk 1 above)
Comment (jeanphilippeD) [inline src/data_handler/idata_op.rs]: Can you do something like:
```rust
             .filter_map(|(sender, state)| match state {
                RpcState::Actioned(response @ Response::Mutation(Err(_))) => Some((*sender, response.clone())),
                _ => None,
            }) 
```

Comment (octol) [reply src/data_handler/idata_op.rs]: Interesting, that's new syntax I learned today :)

File: src/client_handler.rs [hunk 3]
```diff
@@ -107,11 +107,13 @@ impl ClientHandler {
...
             unwrap!(serde_json::to_string(&our_conn_info))
         );
-        if let Ok(connection_info_file) = write_connection_info(&our_conn_info) {
-            println!(
-                "Writing connection info to: {}",
-                connection_info_file.to_string_lossy()
-            );
+        if cfg!(test) {
+            if let Ok(connection_info_file) = write_connection_info(&our_conn_info) {
+                println!(
```
Comment (madadam) [inline src/client_handler.rs]: Just out of curiosity (it's not part of this PR): why do we need this printed out in tests?

Comment (octol) [reply src/client_handler.rs]: It's a bug, I just pushed a fixup that inverts the check :)

**Comment by octol** [comment]
> Looks good, however I would feel more confident if this PR also included tests for the new behaviour.

There is a test that I enabled as part of this PR though https://github.com/maidsafe/safe_vault/pull/814/files#diff-2a2d8cb84eb7f679ed0fdc62e67e9071R2153, you're thinking of something more extensive?

File: src/client_handler.rs [hunk 4]
```diff
@@ -107,11 +107,13 @@ impl ClientHandler {
...
-        if let Ok(connection_info_file) = write_connection_info(&our_conn_info) {
-            println!(
-                "Writing connection info to: {}",
-                connection_info_file.to_string_lossy()
-            );
+        if !cfg!(feature = "mock") {
+            if let Ok(connection_info_file) = write_connection_info(&our_conn_info) {
+                println!(
+                    "Writing connection info to: {}",
+                    connection_info_file.to_string_lossy()
```
[the remark this answers is not in the document]
Comment (octol) [reply src/client_handler.rs]: I'll update the PR

**Comment by madadam** [comment]
> There is a test that I enabled as part of this PR though https://github.com/maidsafe/safe_vault/pull/814/files#diff-2a2d8cb84eb7f679ed0fdc62e67e9071R2153, you're thinking of something more extensive?

Sorry, must have missed that. I think that should be sufficient for now.
